### ISSUE TITLE
`implicitDependencies` dont work when `.gitignore` include the file

### ISSUE THREAD

**Issue by xiejay97** [nrwl/nx#6821]
## Current Behavior

I create custom executor(ts) and add the output file to project graph by `implicitDependencies`.
But it cant work when `.gitignore` include  the output file, so cache never be broken.
![image](https://user-images.githubusercontent.com/64340763/130431051-785803aa-4050-4afa-b22d-3055af71bb22.png)

## Expected Behavior

I think files in `implicitDependencies` should always add to  project graph even they are included in `.gitignore`.

## Steps to Reproduce

This issue may not be prioritized if details are not provided to help us reproduce the issue.

### Environment

nx:12.17.1

**Comment by kylecordes** [comment]
Can confirm it is **very** surprising in any build system for .gitignore to cause a declared dependency to be silently ignored.

**Comment by WonderPanda** [comment]
This seems like a huge gotcha to me as well and I'm struggling to find a way around it.

In my situation I have a generated swagger.json artifact that is built during CI that should be an implicit dependency of a downstream client that consumes the OpenAPI specification. I don't want to check in the swagger.json file and have a whole bunch of unnecessary diffs showing up on developer machines but I *definitely* want changes to this file in CI to bust the cache for the downstream client build.

Is there any workaround for this? If we explicitly list an implicitDependency I think its contents should be hashed and used to bust the cache regardless of whether it is tracked in Git

**Comment by WonderPanda** [comment]
Bumping this, I really like the community could benefit from better control over implict dependencies for code generated artifacts

**Comment by cgrabmann** [comment]
I'm currently also running into this problem.

I'm running some code generation with an external tool (also not a nx-plugin and for reasons I cannot change that easily) and this is then wrapped in an Nx library.   
The generated code is ignored by an `.gitignore`, because there is no reason to check those files into git.

I also tried to explicitly add them in the `.nxignore` by adding `!/generated/files/**` without success.

I'm currently using Nx 12.10.0.

**Comment by junekpavel** [comment]
Hi, are there any updates in this?

Our applications builds are also depending on files generated by external service, whose are ignored by git.

We were quite fine with manually clearing cache when those file changed, but now we starting to use Nx Cloud and this is something we really struggle with. 

All of things mentioned above, we also tried specifying project inputs (https://nx.dev/configuration/projectjson#inputs-&-namedinputs), but without any success.

Or is there anyone who at least figured out a workaround?

**Comment by Fuco1** [comment]
Just bumped into this as well with an `.env` file defining some API URLs which are necessary during build-time webpack bundling.

**Comment by wereHamster** [comment]
I think I ran into something similar as well.

I have a target which generates a GraphQL schema file (from some inputs), that task is non-cacheable so it always executes. That is fine since it takes just a second to run.

Then there is a second target which uses this generated file along with two others to generate some more files. All three files are mentioned in the target inputs, so I was expecting any change to cause a cache bust.

The GraphQL schema file along with the other generated files are in .gitignore, and so the files which are generated using that as input would not be updated. Once I remove the GraphQL schema file from .gitignore, any changes therein cause downstream artefacts to be updated correctly.

**Comment by rofrantz** [comment]
I have a workaround for this problem until NX decides to fix it:

1. add the following file in `tools/cache-file.js`
```
#!/usr/bin/env node

const { existsSync, readFileSync } = require("fs");
const path = require("path");
const chalk = require("chalk");
const { createHash } = require("crypto");

require("yargs").command({
  command: "hash <filepath>",
  describe: `Outputs the hash of the provided file\nUsage: ./${path.basename(
    __filename
  )} hash [filepath]`,
  builder: (yargs) => {
    yargs
      .option("filepath", {
        demandOption: true,
        describe: "Path to the file",
        type: "string",
      })
      .coerce("filepath", (arg) => {
        if (!existsSync(arg)) {
          throw new Error(
            `Provided '${chalk.underline(arg)}' filepath argument doesn't exist`
          );
        } else {
          return readFileSync(arg, "utf8");
        }
      })
      .fail((msg, err, yargs) => {
        console.error(`${chalk.red("[ERROR]")}: ${msg}`);
        console.error(`${yargs.help()}`);
        process.exit(1);
      });
  },
  handler: (argv) => {
    const hashSum = createHash("sha256");
    hashSum.update(argv.filepath);

    console.log(hashSum.digest("hex"));
  },
}).argv;
```

2. Update your `nx.json` file as described below:
```
    "your-task": {
      "inputs": [
        "{workspaceRoot}/.env",
        "{workspaceRoot}/tools/setenv.ts",
        {
          "runtime": "node .\\tools\\cache-file.js hash .env"  // <-- your file that currently gets ignored due to the fact that it's in .gitignore
        }
      ]
    }
```

**Comment by xiejay97** [comment]
> !tools/executors/**/impl.js

@FrozenPandaz I add `!tools/test.js` to `.nxignore` and add `tools/test.js` to `.gitignore`, I run a build task, then change `test.js` and rerun task, but get cache. nx 15.2.1
![image](https://user-images.githubusercontent.com/64340763/203190227-b5a99f90-3374-4cc4-90ff-983ccda762c8.png)
![image](https://user-images.githubusercontent.com/64340763/203190255-dbcc1235-45ad-4a85-b083-9e22908b0eae.png)
![image](https://user-images.githubusercontent.com/64340763/203190693-f820fd15-ffcd-4f44-96f1-094852511ded.png)

**Comment by a777med** [comment]
> Can you all try adding the following to your `.nxignore` file and see if the issue is resolved?
> 
> ```
> !tools/executors/**/impl.js
> ```
> 
> This should negate the `.gitignore` and everything should work as expected but the file will still be `gitignored`

Did not help.

**Comment by a777med** [comment]
Git-ignored files are also not seen when put in `namedInputs`. I think this should be at least mentioned in the docs.

**Comment by y-code** [comment]
I have a similar issue with `inputs`.

I have protobuf source files out of the Nx workspace and generate TypeScript code from them by a script defined in the package.json of an Nx project. With the configurations below, I'm running this task by command `npx nx protoc-gen-ng`.

To make the caching work for the script task, I needed to create a symlink inside the workspace linking to the directory where there are the protobuf files. It was because the caching didn't work with `"inputs": ["{workspaceRoot}/../proto/*.proto"]`. So, I have a _symlink_ `{workspaceRoot}/proto` and configures the inputs as seen below.

Now, the caching works. However, I want to make git ignore the files under the _symlink_, and the problem affects me. Once I add the path to the _symlink_ to _.gitignore_, the caching stops working.

**/nx.json**
```json
{
  "targetDefaults": {
    "protoc-gen-ng": {
      "inputs": ["{workspaceRoot}/proto/*.proto"]
    }
  },
}
```
**/libs/proto-lib/package.json**
```json
{
  "scripts": {
    "protoc-gen-ng": "protoc --plugin=../../node_modules/.bin/protoc-gen-ng --ng_out=./src/lib/generated -I ../../proto ../../proto/*.proto ",
  }
}
```

The idea with `.nxignore` by @FrozenPandaz does not help this case either. But I like that way of configuration. I hope we can have it work.

**Comment by WonderPanda** [comment]
Bumping this as it is definitely still an issue without a clear solution that requires hacky workarounds

**Comment by a777med** [comment]
I am having to commit my .env file to avoid that bug and is not acceptable by our DevOps teams. I always have to explain why to anyone who sees it.

**Comment by wereHamster** [comment]
@a777med I didn't want to include the .env file in the history, so I generate the file as part of the build process, and then delete it afterwards (the deploy script does it automatically). That way I don't have to commit it nor put into .gitignore.

**Comment by simondotm** [comment]
Similar issues happening for me with assets in a `@nrwl/js:tsc` executor.
My pre-build process generates a local json file artefact which is in my `build.options.assets` array.

When this file is in my `.gitignore` (necessarily so), it doesn't get copied to dist when the target is built as I wanted.
When removed from `.gitignore`, it does, but this is not helpful.

**Comment by WonderPanda** [comment]
Bump

**Comment by karol-f** [comment]
Bump

**Comment by andersonba** [comment]
Bump

**Comment by X4V1** [comment]
> Similar issues happening for me with assets in a `@nrwl/js:tsc` executor. My pre-build process generates a local json file artefact which is in my `build.options.assets` array.
> 
> When this file is in my `.gitignore` (necessarily so), it doesn't get copied to dist when the target is built. When removed from `.gitignore`, it does.
> 
> Cannot think of a workaround.

I'm in the exact same situation. Did you find a solution ?

**Comment by X4V1** [comment]
quick answer to myself, nxignore fixed my problem

**Comment by karol-f** [comment]
> quick answer to myself, nxignore fixed my problem

Do you mean this [solution](https://github.com/nrwl/nx/issues/6821#issuecomment-1322761185)?

@a777med wrote that this didn't work for him. Do you have different output?

**Comment by FrozenPandaz** [comment]
Hey all,

Sorry for the delay.

`implicitDependencies` onto files in `nx.json` are no longer recommended. Please use inputs instead: https://nx.dev/more-concepts/customizing-inputs#customizing-inputs-and-named-inputs

.gitignore'd files should be valid inputs. We'll have to investigate why they are not being hashed.

**Comment by halfbakedsneed** [comment]
I managed to get `inputs` considering entries within `.gitignore` by adding a compensating exclusionary entry in `.nxignore`.

**Do note,** if the files you want included in the hash sit within a `.gitignore`d directory entry that **is not** followed by a `/*`, the files **will not** be included in the hash. Even if you have an exclusionary entry in `.nxignore` for the file(s)!

The following configuration ended up working for me:

**.gitignore**

```diff
- .vercel
+ .vercel/*
```

**.nxignore**

```
!.vercel/.env.production.local
```

**project.json**

```json
{
  "targets": {
    "build": {
      "inputs": ["{workspaceRoot}/.vercel/.env.production.local"]
    }
  }
}
```

**Comment by WonderPanda** [comment]
> I managed to get `inputs` to consider entries within `.gitignore` by adding a compensating exclusionary entry in `.nxignore`.
> 
> **Do note,** if the files you want included in the hash sit within a `.gitignore`d directory entry that **is not** followed by a `/*`, the files **will not** be included in the hash. Even if you have an exclusionary entry in `.nxignore` for the file(s)!
> 
> The following configuration ended up working for me:
> 
> **.gitignore**
> 
> ```diff
> - someDir
> + someDir/*
> ```
> 
> **.nxignore**
> 
> ```
> !someDir/.env.production.local
> ```
> 
> **project.json**
> 
> ```json
> {
>   "targets": {
>     "build": {
>       "inputs": ["{workspaceRoot}/someDir/.env.production.local"]
>     }
>   }
> }
> ```

Oh dang that's really interesting! Been watching this issue for ages I hope that it will finally help us simplify our setup

**Comment by karol-f** [comment]
Based on @rofrantz answer, I've modified it a bit to accept globs:

`//generateHash.js`

```js
const fs = require('fs');
const crypto = require('crypto');
const { glob } = require('glob');
const yargs = require('yargs');

const { argv } = yargs
	.command('globs', 'Get globs hash', {
		list: {
			description: 'List',
			type: 'array',
		},
	})
	.help();

const list = argv.list || [];
const files = glob.sync(list);
const hash = crypto.createHash('sha256');

files.forEach((file) => {
	const fileContent = fs.readFileSync(file);
	hash.update(fileContent);
});

console.log(hash.digest('hex'));
```

usage in `nx.json`:

```js
{
	"namedInputs": {
		"default": [
			{ "runtime": "node ./packages/scripts/utilities/calculateHash.js globs --list packages/**/deeplyNestedFile.ts packages/someOtherFile.js" }
		]
	},
        "targetDefaults": {
		"code-quality": {
			"inputs": ["default"]
		},
		"lint": {
			"inputs": ["default"]
		}
        }
}
```

**Comment by FrozenPandaz** [comment]
> There's been a related open issue for a few years: https://github.com/BurntSushi/ripgrep/issues/1050

Yes, this is the underlying crate we use.

So what does work is:

.gitignore
```
dir
```

.nxignore
```
!dir
```

But what does not work is:

.gitignore
```
dir
```

.nxignore
```
!dir/subdir
```

I think we've derailed from the original issue.

`implicitDependencies` has been deprecated now. We narrowed the issue down to a specific pattern of "unignoring" gitignored files/directories

I created a new issue, https://github.com/nrwl/nx/issues/19796, to reflect the new issue and will close this issue though I think it will be blocked on the `ignores` crate unless we can do something else about it.
